### PR TITLE
Support hierarchical dict

### LINKED ISSUE #1144: Add support for hierarchical dict
### Motivation
Since v0.7.0, LightningModule accepts dict hparams, however, still TensorBoardLogger raises an error with hierarchical dict. Considering the compatibility of the other package, especially Hydra #807, hierarchical dict should be accepted by any loggers.

### Pitch
- Flatten hierarchical dict before hparam logging

### Alternatives
The function `_convert_params` in [loggers/base.py](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/loggers/base.py#L45) will be changed like:

```python
def _convert_params(self, params: Union[Dict[str, Any], Namespace]) -> Dict[str, Any]:
    # in case converting from namespace
    if isinstance(params, Namespace):
        params = vars(params)

    # added part
    params = flatten_dict(params)  # convert {'a': {'b': 'c'}} -> {'a/b': 'c'}

    if params is None:
        params = {}

    return params
```

### PR DESCRIPTION
**Pull request by S-aiueo32**
# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure to update the docs?   
- [x] Did you write any new necessary tests?  
- [x] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CHANGELOG.md)?

## What does this PR do?
Fixes #1144.

The main change of this PR is adding the function `_flatten_dict` to `LightningLoggerBase`.
This function flattens hierarchical dictionaries and nested Namespace objects:

```python
# hierarchical dict
_flatten_dict({'a': {'b': 'c'}})  # => {'a/b': 'c'}

# nested Namespace
namespace = Namespace(foo=Namespace(bar='buzz'))
namespace = vars(namespace)  # this conversion for top-level is done in `_convert_params`
_flatten_dict(namespace )  # => {'foo/bar': 'buzz'}
```

Considering the compatibility to all loggers, `_flatten_dict` is called in the `log_hyperparams` function of each logger class that needs flattening:
```python
def log_hyperparams(self, params: Union[Dict[str, Any], Namespace]) -> None:
    params = self._convert_params(params)
    params = self._flatten_dict(params)
```

In other words, this function is not called by the loggers as represented by WanbLogger, which supports hierarchical dictionaries by default, to keep them the hierarchy.

**Note1:** WanbLogger supports hierarchical dictionaries, but not nested Namespace. It does not raise an error, however, the output yaml file on the UI is displayed as:
```yml
foo:
  desc: null
  value: Namespace(bar='buzz')
```

**Note2:** TRAINS also supports hierarchical dictionaries, but the UI destroys the hierarchy as done by `_flatten_dict`. To add support for nested Namespace, `TrainsLogger` calls `_flatten_dict`.

### CODE REVIEW

**Comment by awaelchli** [comment]
very nice!

File: pytorch_lightning/loggers/base.py [hunk 1]
```diff
@@ -55,6 +55,35 @@ def _convert_params(params: Union[Dict[str, Any], Namespace]) -> Dict[str, Any]:
...
+        def _dict_generator(_dict, prefixes=None):
+            prefixes = prefixes[:] if prefixes else []
+            if isinstance(_dict, dict):
+                for key, value in _dict.items():
+                    if isinstance(value, dict):
+                        for d in _dict_generator(value, prefixes + [key]):
+                            yield d
+                    elif isinstance(value, Namespace):
+                        for d in _dict_generator(vars(value), prefixes + [key]):
+                            yield d
```
Comment (Borda) [inline pytorch_lightning/loggers/base.py]:
```suggestion
                    if isinstance(value, (dict, Namespace)):
                        value = vars(value) if isinstance(value, Namespace) else value
                        for d in _dict_generator(value, prefixes + [key]):
                            yield d
```

File: pytorch_lightning/loggers/base.py [hunk 2]
```diff
@@ -55,6 +55,35 @@ def _convert_params(params: Union[Dict[str, Any], Namespace]) -> Dict[str, Any]:
...
+
+        Args:
+            params: Dictionary contains hparams
+            delimiter: Delimiter to express the hierarchy. Defaults to '/'.
+
+        Returns:
+            Flatten dict.
+        """
+
+        def _dict_generator(_dict, prefixes=None):
```
Comment (Borda) [inline pytorch_lightning/loggers/base.py]: what about `_dict` >> `input_dict`

File: pytorch_lightning/loggers/base.py [hunk 3]
```diff
@@ -55,6 +55,35 @@ def _convert_params(params: Union[Dict[str, Any], Namespace]) -> Dict[str, Any]:
...
+    def _flatten_dict(params: Dict[str, Any], delimiter: str = '/') -> Dict[str, Any]:
+        """Flatten hierarchical dict e.g. {'a': {'b': 'c'}} -> {'a/b': 'c'}.
+
+        Args:
+            params: Dictionary contains hparams
+            delimiter: Delimiter to express the hierarchy. Defaults to '/'.
+
+        Returns:
+            Flatten dict.
+        """
```
Comment (Borda) [inline pytorch_lightning/loggers/base.py]: rather add doctest
```
>>> LightningLoggerBase.flatten_dict({'a': {'b': 'c'}})
{'a/b': 'c'}
>>> LightningLoggerBase.flatten_dict({'a': Namespace(b= 123)})
{'a/b': 123}
```

File: CHANGELOG.md [hunk 4]
```diff
@@ -8,6 +8,7 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 
 ### Added
 
+- Add support for hierarchical dict ([#1152](https://github.com/PyTorchLightning/pytorch-lightning/pull/1152))
```
Comment (Borda) [inline CHANGELOG.md]:
```suggestion
- Added support for hierarchical dict ([#1152](https://github.com/PyTorchLightning/pytorch-lightning/pull/1152))
```

File: pytorch_lightning/loggers/base.py (hunk 3 above)
Comment (Borda) [inline pytorch_lightning/loggers/base.py]:
```suggestion
        Examples:
            >>> LightningLoggerBase.flatten_dict({'a': {'b': 'c'}})
            {'a/b': 'c'}
            >>> LightningLoggerBase.flatten_dict({'a': Namespace(b= 123)})
            {'a/b': 123}
        """
```

**Comment by KeAWang** [comment]
Is there a reason that `wandb.py` wasn't changed to flatten dict?

**Comment by awaelchli** [comment]
probably got forgotten, is already done, or not necessary. Did you check, is it necessary to do it @KeAWang ?

**Comment by awaelchli** [comment]
See here: https://docs.wandb.com/library/config
They have thought of it already :)
